1-intro/intermediate.ipynb

```diff
@@ -0,0 +1,354 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "3ac6f557-376a-42e1-a92b-e6a83aceaf04",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "fruits = ['apple', 'banana', 'cherry']\n",
+    "for index, fruit in enumerate(fruits):\n",
+    "    print(f'Index {index}: {fruit}')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e57c9b49-14a4-49ba-8055-631a3a25d6b9",
+   "metadata": {},
+   "source": [
+    "### Generators\n",
+    "Generators are a way to create iterators in Python. Generators allow you to generate values on the fly instead of up-front (called lazy evaluation) which is very useful when dealing with large datasets or when you want to conserve memory.\n",
+    "\n",
+    "Generators are typically created using a special type of function called a generator function. These functions use the `yield` keyword to give values one at a time, rather than returning a single result. When a generator function is called, it doesn't execute the entire function at once; it pauses at each `yield` statement until you request the next value.\n",
+    "\n",
+    "Here is a very basic example:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "823ddf43-8492-4fc1-844e-e2002363721e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def my_generator():\n",
+    "    yield \"a\"\n",
+    "    yield \"b\"\n",
+    "    yield \"c\"\n",
+    "\n",
+    "gen = my_generator()\n",
+    "for value in gen:\n",
+    "    print(value)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d1d53271-1feb-4b32-ae41-a82194c01ede",
+   "metadata": {},
+   "source": [
+    "The key feature of generators is that they preserve their internal state. When you iterate over a generator and then continue later, it remembers where it left off."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "d2127995-e66b-4733-8882-960c9b87caea",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "gen = my_generator()  # Redefine the generator\n",
+    "print(next(gen))\n",
+    "print(next(gen))\n",
+    "print(next(gen))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "edf5e02e-0bb5-4daf-ad9e-267754ca211a",
+   "metadata": {},
+   "source": [
+    "The combination of lazy evaluation and state perservation allows you to make generators that are practically infinate in size. Here is one that can generate an infinate sequence of natural numbers."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "944be3ce-a1dd-4b07-99c7-c8126f1d8e5e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def natural_numbers():\n",
+    "    num = 1\n",
+    "    while True:\n",
+    "        yield num\n",
+    "        num += 1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "381ce226-2e74-4813-9f12-26d96b8480ce",
+   "metadata": {},
+   "source": [
+    "### The unpacking operator\n",
```

Comment:
The unpacking operator is used in a solution but it is never explained, and if we don't want to devote more real estate in the manual to it, then it belongs here.

Reply:
Iterable unpacking is not useful for the courses, so I wouldn't spend any time on it. The one-liner exercise solution can be replaced with an even shorter one-liner:
```python
print("\n".join(min(i, 10 - i) * "*" for i in range(1, 10)))
```
The closely related concept of accepting and passing on `**kwargs` can be useful for writing plotting functions because it allows passing on arguments to the underlying `matplotlib` function without having to write a very verbose function signature. Such use of `**kwargs` is demonstrated in the Dynamical Astronomy notebooks in a clear example. The reasoning for having it there and not in the Python introduction materials was that spreading the lessons out in time makes them more accessible and that most of the students taking the Python introduction will also take Dynamical Astronomy anyways. Using `**kwargs` could be brought up in the `matplotlib` lesson, if there is time for it.